homeworks/03-data-structure/binary-tree/binaryTree.ts

```diff
@@ -0,0 +1,208 @@
+    }
+
+
+    public constructor() {
+        this.size = 0
+        this.root = null
+    }
+
+    public currentSize(): number {
+        return this.size
+    }
+
+    public empty(): boolean {
+        return this.size === 0
+    }
+
+    public has(checkValue: ValueType): boolean {
+        return this.findElement(checkValue) !== null
+    }
+
+    public insert(newValue: ValueType): void {
+        if (this.has(newValue))
+            return
+
+        if (this.size !== 0) {
+            let element: BinaryTreeNode<ValueType> = this.root,
+                parent: BinaryTreeNode<ValueType> = this.root
+
+            while (element != null) {
+                parent = element;
+                element = newValue <= element.value ? element.leftChild : element.rightChild
+            }
+
+            const newNode: BinaryTreeNode<ValueType> = new BinaryTreeNode(newValue, parent)
+            if (newValue <= parent.value) {
+                parent.leftChild = newNode
+                newNode.columnNumber = parent.columnNumber - 1
+            } else {
+                parent.rightChild = newNode
+                newNode.columnNumber = parent.columnNumber + 1
+            }
+        } else {
+            this.root = new BinaryTreeNode(newValue, null)
+        }
+
+        ++this.size
+    }
+
+    public remove(deletedValue: ValueType): void {
+        const searchElement: BinaryTreeNode<ValueType> = this.findElement(deletedValue)
+
+        if (searchElement !== null && this.size > 0) {
+            this.deleteByPointer(searchElement)
+            --this.size
+            this.updateColumnNumbers()
+        }
+
+        if (this.size === 0)
+            this.root = null
+    }
+
+    public traverse(orderType: TraverseType): ValueType[] {
+        const response: ValueType[] = []
+
+        switch (orderType) {
+            case TraverseType.BreadthFirst:
+                this.root !== null && this.breadthFirstTraverse(this.root, response)
+                break
+            case TraverseType.PreOrder:
+                this.root !== null && this.recursivePreOrderTraverse(this.root, response)
+                break
+            case TraverseType.PostOrder:
+                this.root !== null && this.recursivePostOrderTraverse(this.root, response)
+                break
+            case TraverseType.InOrder:
+                this.root !== null && this.recursiveInOrderTraverse(this.root, response, () => true)
+                break
+        }
+
+        return response
+    }
+
+    public getColumn(column: number): ValueType[] {
+        const response: ValueType[] = []
+        this.recursiveInOrderTraverse(this.root, response, (node: BinaryTreeNode<ValueType>) => node.columnNumber === column)
+        return response
+    }
+
+    public clear(): void {
+        this.root !== null && this.recursiveClear(this.root)
+        this.size = 0
+        this.root = null
+    }
+}
+
+class BinarySearchTree extends BinaryTree<number> {
```

Comment:
You don't have a separate interface as asked in the task.
Overall, this class looks useless in this case.